pkg/controller/apmserver/services_test.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package apmserver
+
+import (
+	"testing"
+
+	"github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1beta1"
+	commonv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1beta1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/apmserver/labels"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/compare"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestNewService(t *testing.T) {
+	testCases := []struct {
+		name     string
+		httpConf commonv1beta1.HTTPConfig
+		wantSvc  func() corev1.Service
+	}{
+		{
+			name: "no TLS",
+			httpConf: commonv1beta1.HTTPConfig{
+				TLS: commonv1beta1.TLSOptions{
+					SelfSignedCertificate: &commonv1beta1.SelfSignedCertificate{
+						Disabled: true,
+					},
+				},
+			},
+			wantSvc: mkService,
+		},
+		{
+			name: "self-signed certificate",
+			httpConf: commonv1beta1.HTTPConfig{
+				TLS: commonv1beta1.TLSOptions{
+					SelfSignedCertificate: &commonv1beta1.SelfSignedCertificate{
+						SubjectAlternativeNames: []commonv1beta1.SubjectAlternativeName{
+							{
+								DNS: "apm-test.local",
+							},
+						},
+					},
+				},
+			},
+			wantSvc: func() corev1.Service {
+				svc := mkService()
+				svc.Spec.Ports[0].Name = "https"
+				return svc
+			},
+		},
+		{
+			name: "user-provided certificate",
+			httpConf: commonv1beta1.HTTPConfig{
+				TLS: commonv1beta1.TLSOptions{
+					Certificate: commonv1beta1.SecretRef{
+						SecretName: "my-cert",
+					},
+				},
+			},
+			wantSvc: func() corev1.Service {
+				svc := mkService()
+				svc.Spec.Ports[0].Name = "https"
+				return svc
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			apm := mkAPMServer(tc.httpConf)
+			haveSvc := NewService(apm)
+			compare.JSONEqual(t, tc.wantSvc(), haveSvc)
```

Comment:
what does comparing marshalled json get us that comparing structs does not?

Reply:
I guess we have several options:
- `reflect.DeepEqual()`
- string representation comparison through json marshalling (what this PR does)
- `hash.Object()` comparison (that, behind the scenes, relies on a string representation of the object through spew)
- https://github.com/google/go-cmp

In unit tests I don't think it matters much, even though performance may vary 🤷‍♂  (I would personally have used `reflect.DeepEqual()` since builtin but it may have some corner cases I don't know about?)

Reply:
The main reason is that `reflect.DeepEqual` and `require.Equal` produce diffs that are very difficult to read when there are lots of nested elements. JSON diff is much nicer because the output actually shows you where the problem is. Since all three require reflection (the heaviest part) I think the tiny overhead of JSON conversion is acceptable in a unit testing scenario.

I would love to use `go-cmp` but decided against adding an extra dependency for just this test.

Reply:
Thanks for the details 👍 
We also use "github.com/go-test/deep" in other parts of the code. Do you know how does it compare to the json conversion? I feel like we should settle on one and use it everywhere.